### PR TITLE
Add WT_RST_STREAM and WT_STOP_SENDING

### PR DESCRIPTION
**Pull request by afrind**

### CODE REVIEW

File: draft-ietf-webtrans-http2.md [hunk 1]
```diff
@@ -269,6 +269,116 @@ exist, is not a stream established via extended CONNECT to use the
...
+ 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
++---------------------------------------------------------------+
+|                        Error Code (32)                        |
++---------------------------------------------------------------+
+~~~
+{: #fig-wt_stop_sending title="WT_STOP_SENDING Frame Format"}
+
+The WT_STOP_SENDING frame contains a single unsigned, 32-bit integer identifying
+the error code (Section 7).  The error code indicates why the reading the stream
+is being abanonded.
```
Comment (ekinnear) [inline draft-ietf-webtrans-http2.md:343]:
```suggestion
is being abandoned.
```

**Review by ekinnear — APPROVED** [review]
Thanks @afrind!

File: draft-ietf-webtrans-http2.md [hunk 2]
```diff
@@ -269,6 +269,116 @@ exist, is not a stream established via extended CONNECT to use the
...
+WT_STOP_SENDING frames MUST be associated with a WebTransport stream.  If a
+WT_STOP_SENDING frame is received with a stream identifier of 0x0, or a request
+or push stream, the recipient MUST treat this as a connection error of type
+PROTOCOL_ERROR.
+
+WT_STOP_SENDING frames MUST NOT be sent for a stream in any state other than
+"open" or "half-closed (local)".  It is possible to receive a WT_STOP_SENDING in
+another state however, because the sender might have closed or reset the stream
+while the WT_STOP_SENDING was in flight.  If a WT_STOP_SENDING frame identifying
+a stream that is already "closed" or "half-closed (local"), the recipient SHOULD
```
Comment (martinthomson) [inline draft-ietf-webtrans-http2.md:370]:
```suggestion
a stream that is already "closed" or "half-closed (local)", the recipient SHOULD
```

File: draft-ietf-webtrans-http2.md [hunk 3]
```diff
@@ -269,6 +269,116 @@ exist, is not a stream established via extended CONNECT to use the
...
+ignore the frame.
+
+It is also possible to receive DATA frames on a WebTransport stream in the
+"half-closed (remote)" or "closed" states if the stream transitioned there via
+WT_STOP_SENDING.  If a DATA frame is received on a WebTransport stream in one of
+these states, the recipient MUST ignore the frame and credit the frame length to
```
Comment (martinthomson) [inline draft-ietf-webtrans-http2.md:376]: Is "credit" the right term here?  "count the frame length toward the FC window" perhaps.

**Review by martinthomson — COMMENTED** [review]
Thanks for writing this up Alan.

This works.  Though I'm not sure about the STOP_SENDING analogue; QUIC permits the sender to choose the ultimate code for RESET_STREAM, even if it is generally copied  from STOP_SENDING.  I know that we mostly don't need to include RESET_STREAM here as the stream size accounting isn't as critical, so maybe perfect fidelity isn't needed.

**Comment by afrind** [comment]
Thanks for the review Martin.  I suppose the higher fidelity version of STOP_SENDING would not change the state of the stream at all, but induce the peer to send a RST_STREAM as QUIC does.  Maybe that is better.

**Comment by martinthomson** [comment]
Actually, there is worse: when I send STOP_SENDING, I no longer know when the peer has received and acted on that without an explicit stream termination signal, so I can't transition the stream to a closed state.

**Comment by afrind** [comment]
> Actually, there is worse: when I send STOP_SENDING, I no longer know when the peer has received and acted on that without an explicit stream termination signal, so I can't transition the stream to a closed state.

Wait, how is this different than QUIC?  You can't close a bidirectional stream until the peer sends you a FIN or RST.

**Comment by martinthomson** [comment]
Right; I was reading your description of STOP_SENDING as changing stream state, either at the point that it was sent or received.  In that case, if a stream is "open", the sender of STOP_SENDING might believe that all it needs to do is send RESET_STREAM to close the stream completely.  If it does so, the peer will receive those frames and then both peers will believe the stream to be closed.  But the sender of STOP_SENDING has no way to distinguish between that set of circumstances and one where the peer takes an arbitrary amount of time to process the frames.  The receiver of STOP_SENDING might just keep sending data indefinitely.  How does the receiver of those frames know when the other side has truly stopped?

You might say "too bad".  After all, the only thing you need to do is keep counting against connection-level flow control, but it makes the "closed" state that much harder to manage.

**Comment by afrind** [comment]
As written, STOP_SENDING does change stream state, which would allow the sender to move to closed unilaterally (for example, send WT_RST_STREAM and WT_STOP_SENDING together).  DATA frames may trickle in after that for the closed stream, you just keep handling them for connection flow control -- this is the same as RST_STREAM in vanilla H2.

If we make it like QUIC instead, where STOP_SENDING does not affect state, the sender of the STOP_SENDING has to wait for WT_RST_STREAM or a FIN to move to closed.  

So I think either model is fine and I don't have a stylistic preference.  I'm wondering if it has an effect on the stream limit accounting, but I don't think it does?  The H2 model frees the streams 1/2 RTT faster.

**Comment by martinthomson** [comment]
Ugh, I should know this stuff better, I only recently rewrote the closing bit in h2 to deal with this problem.

> this is the same as RST_STREAM in vanilla H2.

Of course.  It's bad, but no worse than status quo.  Thanks for correcting me.

(Full transparency: I think that the QUIC stream model is so much better than h2 that I can no longer stomach these minor infidelities in h2.  This PR does manage to balance the need to provide API consistency across versions without fixing/changing h2's problems.  I'm sure I'll manage to swallow my distaste eventually if we go this way.)

**Comment by afrind** [comment]
> I think that the QUIC stream model is so much better than h2 that I can no longer stomach these minor infidelities in h2

Heh. Since this is merging an H3 feature into H2 I feel like we could pick either model and still claim to be consistent.  For now, I'll leave this H2 style.

**Pull request merged**